{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d7f0f5dd-1cb0-4f9b-a35b-ba348a131cef",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/qukun/lyz32/miniconda3/envs/scarches/lib/python3.9/site-packages/umap/distances.py:1063: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n",
      "  @numba.jit()\n",
      "/home/qukun/lyz32/miniconda3/envs/scarches/lib/python3.9/site-packages/umap/distances.py:1071: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n",
      "  @numba.jit()\n",
      "/home/qukun/lyz32/miniconda3/envs/scarches/lib/python3.9/site-packages/umap/distances.py:1086: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n",
      "  @numba.jit()\n",
      "/home/qukun/lyz32/miniconda3/envs/scarches/lib/python3.9/site-packages/umap/umap_.py:660: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n",
      "  @numba.jit()\n",
      "/home/qukun/lyz32/miniconda3/envs/scarches/lib/python3.9/site-packages/umap/__init__.py:9: ImportWarning: Tensorflow not installed; ParametricUMAP will be unavailable\n",
      "  warn(\n",
      "WARNING:root:In order to use the mouse gastrulation seqFISH datsets, please install squidpy (see https://github.com/scverse/squidpy).\n",
      "/home/qukun/lyz32/miniconda3/envs/scarches/lib/python3.9/site-packages/lightning/fabric/__init__.py:36: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('lightning.fabric')`.\n",
      "Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages\n",
      "  __import__(\"pkg_resources\").declare_namespace(__name__)\n",
      "/home/qukun/lyz32/miniconda3/envs/scarches/lib/python3.9/site-packages/pkg_resources/__init__.py:2350: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('lightning')`.\n",
      "Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages\n",
      "  declare_namespace(parent)\n",
      "/home/qukun/lyz32/miniconda3/envs/scarches/lib/python3.9/site-packages/torchmetrics/utilities/imports.py:24: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.\n",
      "  _PYTHON_LOWER_3_8 = LooseVersion(_PYTHON_VERSION) < LooseVersion(\"3.8\")\n",
      "/home/qukun/lyz32/miniconda3/envs/scarches/lib/python3.9/site-packages/lightning/pytorch/__init__.py:36: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('lightning.pytorch')`.\n",
      "Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages\n",
      "  __import__(\"pkg_resources\").declare_namespace(__name__)\n",
      "/home/qukun/lyz32/miniconda3/envs/scarches/lib/python3.9/site-packages/pkg_resources/__init__.py:2350: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('lightning')`.\n",
      "Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages\n",
      "  declare_namespace(parent)\n",
      "/home/qukun/lyz32/miniconda3/envs/scarches/lib/python3.9/site-packages/scvi/_settings.py:63: UserWarning: Since v1.0.0, scvi-tools no longer uses a random seed by default. Run `scvi.settings.seed = 0` to reproduce results from previous versions.\n",
      "  self.seed = seed\n",
      "/home/qukun/lyz32/miniconda3/envs/scarches/lib/python3.9/site-packages/scvi/_settings.py:70: UserWarning: Setting `dl_pin_memory_gpu_training` is deprecated in v1.0 and will be removed in v1.1. Please pass in `pin_memory` to the data loaders instead.\n",
      "  self.dl_pin_memory_gpu_training = (\n",
      "WARNING:root:In order to use sagenet models, please install pytorch geometric (see https://pytorch-geometric.readthedocs.io) and \n",
      " captum (see https://github.com/pytorch/captum).\n",
      "WARNING:root:mvTCR is not installed. To use mvTCR models, please install it first using \"pip install mvtcr\"\n",
      "/home/qukun/lyz32/miniconda3/envs/scarches/lib/python3.9/site-packages/scanpy/_settings.py:450: DeprecationWarning: `set_matplotlib_formats` is deprecated since IPython 7.23, directly use `matplotlib_inline.backend_inline.set_matplotlib_formats()`\n",
      "  IPython.display.set_matplotlib_formats(*ipython_format)\n"
     ]
    }
   ],
   "source": [
    "import scib\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "from multiprocessing import Pool \n",
    "from scipy.io import mmread\n",
    "from scipy.sparse import csr_matrix\n",
    "import muon\n",
    "import scarches as sca\n",
    "import scanpy as sc\n",
    "from scib_metrics.benchmark import Benchmarker\n",
    "import scib_metrics\n",
    "from typing import Optional\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31d35c65-0f24-43d9-a276-d74983a89773",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dt_list = ['3_brain_SNARE',\n",
    "#          '4_brain_ISSAAC_seq',\n",
    "#          '5_human_brain_10x',\n",
    "#          '6_pbmc_granulocyte_sorted_10k',\n",
    "#          '7_human_pbmc_10x',\n",
    "#          '8_pbmc_unsorted_10k',\n",
    "#          '9_pbmc_unsorted_3k',\n",
    "#          '10_GSE201402_down',\n",
    "#            '11_GSE194122',\n",
    "#          '12_GSE156478_ctrl',\n",
    "#          '12_GSE156478_stim']\n",
    "\n",
    "dt_list = ['10_GSE201402_down']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f93c88ed-621a-4799-896a-de50c7ce0d8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "for dataset in dt_list:\n",
    "    latent = pd.DataFrame(np.load('../Integration/results/Vertical/RNA_ATAC/' + dataset + '/scMVP.npy'))\n",
    "    latent.to_csv('../Integration/results/Vertical/RNA_ATAC/' + dataset + '/scMVP.csv', header = None, index=False)\n",
    "    # latent = pd.read_csv('../Integration/results/Vertical/RNA_ATAC/' + dataset + '/scAI.csv', header = None)\n",
    "    # latent.T.to_csv('../Integration/results/Vertical/RNA_ATAC/' + dataset + '/scAI.csv', header = None, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "76c1ba2a-9422-4646-81db-f0ed1d97bb72",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_RNA_ATAC(RNA_path,ATAC_path):\n",
    "    # gene expression\n",
    "    cell_names = pd.read_csv(RNA_path+'barcodes.tsv', sep = '\\t', header=None, index_col=None)\n",
    "    cell_names.columns =  ['cell_ids'] \n",
    "    cell_names['cell_ids'] = cell_names['cell_ids'].str.replace('.','-')\n",
    "    X = csr_matrix(mmread(RNA_path+'matrix.mtx').T)\n",
    "    gene_names = pd.read_csv(RNA_path+'features.tsv', sep = '\\t',  header=None, index_col=None) \n",
    "    gene_names.columns =  ['gene_ids'] \n",
    "    adata_RNA = ad.AnnData(X, obs=pd.DataFrame(index=cell_names.cell_ids), var=pd.DataFrame(index = gene_names.gene_ids))\n",
    "    adata_RNA.var_names_make_unique()\n",
    "    # peak information\n",
    "    cell_names = pd.read_csv(ATAC_path + 'barcodes.tsv', sep = '\\t', header=None, index_col=None)\n",
    "    cell_names.columns =  ['cell_ids'] \n",
    "    cell_names['cell_ids'] = cell_names['cell_ids'].str.replace('.','-')\n",
    "    X = csr_matrix(mmread(ATAC_path + 'matrix.mtx').T)\n",
    "    peak_name = pd.read_csv(ATAC_path + 'features.tsv', sep = '\\t',header=None,index_col=None)\n",
    "    peak_name.columns = ['peak_ids']\n",
    "    adata_ATAC  = ad.AnnData(X, obs=pd.DataFrame(index=cell_names.cell_ids), var=pd.DataFrame(index = peak_name.peak_ids))\n",
    "    return adata_RNA, adata_ATAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "70e6c220-519d-4f59-a248-e948e219c7a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_metrics(dataset):\n",
    "    data_path = '../Integration/dataset/Vertical/RNA_ATAC/' + dataset\n",
    "    adata_RNA, adata_ATAC = read_RNA_ATAC(data_path+\"/RNA/\",data_path+\"/ATAC/\")\n",
    "\n",
    "    adata_RNA.layers[\"counts\"] = adata_RNA.X.copy()\n",
    "    sc.pp.normalize_total(adata_RNA)\n",
    "    sc.pp.log1p(adata_RNA)\n",
    "    sc.pp.highly_variable_genes(\n",
    "        adata_RNA,\n",
    "        flavor=\"seurat_v3\",\n",
    "        n_top_genes=4000,\n",
    "        subset=False\n",
    "    )\n",
    "    adata_RNA = adata_RNA[:, adata_RNA.var.highly_variable].copy()\n",
    "\n",
    "    adata_ATAC.layers['counts'] = adata_ATAC.X.copy()\n",
    "    sc.pp.normalize_total(adata_ATAC, target_sum=1e4)\n",
    "    sc.pp.log1p(adata_ATAC)\n",
    "    adata_ATAC.layers['log-norm'] = adata_ATAC.X.copy()\n",
    "    sc.pp.highly_variable_genes(adata_ATAC, n_top_genes=30000)\n",
    "    adata_ATAC = adata_ATAC[:, adata_ATAC.var.highly_variable].copy()\n",
    "    \n",
    "    adata = sca.models.organize_multiome_anndatas(\n",
    "        adatas = [[adata_RNA], [adata_ATAC]],    # a list of anndata objects per modality, RNA-seq always goes first\n",
    "        layers = [['counts'], ['log-norm']], # if need to use data from .layers, if None use .X\n",
    "    )\n",
    "\n",
    "    metadata = pd.read_csv(data_path + \"/meta_data.csv\")\n",
    "    metadata['celltype'].index = adata.obs_names\n",
    "    adata.obs['cell_type'] = metadata['celltype'].astype('category')\n",
    "    if np.where(adata.obs[\"cell_type\"].isna())[0].shape[0]!=0:\n",
    "        adata.obs[\"cell_type\"] = adata.obs[\"cell_type\"].cat.add_categories(['NaN'])\n",
    "        adata.obs[\"cell_type\"][np.where(adata.obs[\"cell_type\"].isna())[0]] = 'NaN'\n",
    "    adata.obs['batch'] = ['batch1'] * adata.shape[0]\n",
    "    \n",
    "    result = pd.DataFrame()\n",
    "    metrics_list = []\n",
    "    method_list = ['Multigrate','MultiVI','MOFA','SCOIT','MIRA','scMVP','scVAEIT','MOJITOO','DeepMaps','Schema','scAI']\n",
    "    for method in method_list:\n",
    "        if os.path.exists('../Integration/dataset/Vertical/RNA_ATAC/'+ dataset + '/'+ method + '.csv'):\n",
    "            latent = pd.read_csv('../Integration/dataset/Vertical/RNA_ATAC/'+ dataset + '/'+ method + '.csv', header = None)\n",
    "            latent.index = adata.obs_names\n",
    "            adata.obsm[method] = latent\n",
    "            sc.pp.neighbors(adata, use_rep=method)\n",
    "            sc.tl.umap(adata)\n",
    "            sc.tl.leiden(adata, key_added=\"cluster\")\n",
    "            scib.metrics.cluster_optimal_resolution(adata, cluster_key=\"cluster\", label_key=\"cell_type\")\n",
    "            ari = scib.metrics.ari(adata, cluster_key=\"cluster\", label_key=\"cell_type\")\n",
    "            iso_asw = scib.metrics.isolated_labels_asw(adata, label_key=\"cell_type\", batch_key='batch', embed=method,  verbose = False)\n",
    "            nmi = scib.metrics.nmi(adata, cluster_key=\"cluster\", label_key=\"cell_type\")\n",
    "            clisi = scib.metrics.clisi_graph(adata, label_key=\"cell_type\",use_rep=method, type_='embed')\n",
    "            sht = scib.metrics.silhouette(adata, label_key=\"cell_type\", embed=method, metric='euclidean', scale=True)\n",
    "            metrics_list.append([ari, iso_asw, nmi, clisi, sht, method])\n",
    "            \n",
    "    con = mmread('../Integration/dataset/Vertical/RNA_ATAC/'+ dataset + '/'+ method + '_connectivities.mtx')\n",
    "    dis = mmread('../Integration/dataset/Vertical/RNA_ATAC/'+ dataset + '/'+ method + '_distance.mtx')\n",
    "    adata.uns['neighbors'] = {'connectivities_key': 'connectivities', 'distances_key': 'distances', \n",
    "                              'params': {'n_neighbors': 20, 'method': 'umap', 'random_state': 0, \n",
    "                                         'metric': 'euclidean'}}\n",
    "    adata.uns['neighbors']['distance'] = csr_matrix(dis)\n",
    "    adata.uns['neighbors']['connectivities'] = csr_matrix(con)\n",
    "    adata.obsp['distance'] = csr_matrix(dis)\n",
    "    adata.obsp['connectivities'] = csr_matrix(con)\n",
    "    sc.tl.umap(adata, n_components=20)\n",
    "    scib.metrics.cluster_optimal_resolution(adata, cluster_key=\"cluster\", label_key=\"cell_type\")\n",
    "    ari = scib.metrics.ari(adata, cluster_key=\"cluster\", label_key=\"cell_type\")\n",
    "    iso_asw = scib.metrics.isolated_labels_asw(adata, label_key=\"cell_type\", batch_key='batch', embed=method,  verbose = False)\n",
    "    nmi = scib.metrics.nmi(adata, cluster_key=\"cluster\", label_key=\"cell_type\")\n",
    "    clisi = scib.metrics.clisi_graph(adata, label_key=\"cell_type\",use_rep=method, type_='embed')\n",
    "    sht = scib.metrics.silhouette(adata, label_key=\"cell_type\", embed=method, metric='euclidean', scale=True)\n",
    "    metrics_list.append([ari, iso_asw, nmi, clisi, sht, 'Seurat'])\n",
    "\n",
    "    df = pd.DataFrame(metrics_list,columns = ['Isolated_Labels','method'])\n",
    "    result = pd.merge(result,df, on='method', how='outer')\n",
    "    result['Dataset'] = dataset\n",
    "    result.to_csv(data_path + \"/metrics_result.csv\",index = False)\n",
    "    print(dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1bb9cc69-2be1-4d16-be3d-309e4d9f12a6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: You’re trying to run this on 34000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n",
      "3_brain_SNARE\n",
      "WARNING: You’re trying to run this on 34000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n",
      "4_brain_ISSAAC_seq\n",
      "WARNING: You’re trying to run this on 34000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n",
      "5_human_brain_10x\n",
      "WARNING: You’re trying to run this on 34000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n",
      "6_pbmc_granulocyte_sorted_10k\n",
      "WARNING: You’re trying to run this on 34000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n",
      "7_human_pbmc_10x\n",
      "WARNING: You’re trying to run this on 34000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n",
      "8_pbmc_unsorted_10k\n",
      "WARNING: You’re trying to run this on 34000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n",
      "9_pbmc_unsorted_3k\n",
      "WARNING: You’re trying to run this on 34000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n",
      "10_GSE201402_down\n",
      "WARNING: You’re trying to run this on 34000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n",
      "12_GSE156478_ctrl\n",
      "WARNING: You’re trying to run this on 34000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n",
      "12_GSE156478_stim\n"
     ]
    }
   ],
   "source": [
    "for dataset in dt_list:\n",
    "    count_metrics(dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e323155-0ec1-454f-be01-caeb1f0c06dc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
