# Benchmarking of algorithms predicting protein abundance and chromatin accessibility from single-cell transcriptomics
![WorkFolw](./fig/pipeline.jpg)

__Implementation description__

  we collected 24 RNA+Protein multiomics datasets generated by CITE-seq and REAP-seq, 10 RNA+ATAC multiomics datasets generated by 5 single-cell multiomics sequencing technologies (10x Multiome, SNARE-seq, Paired-seq, SHARE-seq, and ISSAAC-seq), and one RNA+ATAC+Protein single-cell multiomics dataset obtained by DOGMA-seq to 1) systematically evaluate the performance of 12 algorithms for predicting missing protein abundance or chromatin accessibility information by using gene expression. 2) We also explore the impact of batch effects on the accuracy of these algorithms. 3ï¼‰Finally, we evaluate the computational resources consumed by each method.

  We provide example guidance to help researchers select optimal modulity prediction algorithms for working with their datasets:
  the [doc/Tutorial.pdf](https://github.com/QuKunLab/SpatialBenchmarking/blob/main/doc/Tutorial.pdf) is an example showing how to use them to predict protein abundance or chromatin accessibility information by using gene expression.


__Dependencies and requirements for Predicting protein abundance or chromatin accessibility information of cells in scRNA-seq data__

 Before you run the pipeline, please make sure that you have installed and python3, R(4.1) and all the 12 packages(totalVI, scArches, cTP-net, Babel, CMAE, sciPENN, Guanlab-dengkw, scMoGCN, MultiVi, LS_Lab, Seurat and LIGER) :
1. Before the installation of these packages, please install Miniconda to manage all needed software and dependencies. You can download Miniconda from https://conda.io/miniconda.html.
2. Download MultiomeBenchmarking.zip from https://github.com/QuKunLab/MultiomeBenchmarking. Unzipping this package and you will see Benchmarkingenvironment.yml and Config.env.sh located in its folder.
3. Build isolated environment for MultiomeBenchmarking: 
`conda env create -f Benchmarkingenvironment.yml`
4. Activate Benchmarking environment:
`conda activate Benchmarking`
5. `sh Config.env.sh`
6. Enter R and install required packages by command : `install.packages(c('vctrs','rlang','htmlwidgets'))`

Installation of Benchmarking may take about 7-15 minutes to install the dependencies.

__Dependencies and requirements for Predicting chromatin accessibility information or protein abundance of cells in scRNA-seq__

Before you run the pipeline, please make sure that you have installed and python3, R and all the ten packages: Cell2location(Version 0.6a0), RCTD(Version 1.2.0), SpatialDWLS(by Giotto of Version 1.0.4), Stereoscope(within the scvi-tools Version 0.11.0), SPOTlight(Version 0.1.7), Tangram(Version 1.0.0), Seurat(Version 4.0.5), STRIDE(Version 0.0.1b), DestVI(scvi-tools Version 0.11.0), DSTG(Version 0.0.1)
 
 The package has been tested on Linux system (CentOS) and should work in any valid python environment. 

__Tutorial__

  If you want to analysis your own data, the [doc/Tutorial.ipynb](https://github.com/QuKunLab/SpatialBenchmarking/blob/main/doc/Tutorial.ipynb) is an example showing how to use them to predict new spatial gene patterns and cell locations.

  You also can run the jupyter notebook of `BLAST_GenePrediction.ipynb` and `BLAST_CelltypeDeconvolution.ipynb` to reproduce the results of figure2&4 in our paper.
  
  For more details, please see the `SpatialGenes.py` & `Deconvolution.py` in Benchmarking directory.

__Datasets__

  All datasets used are publicly available data, for convenience datasets can be downloaded from: https://drive.google.com/drive/folders/153GW_iJ2tViOfKayJuaAIb8a7op8EWWt?usp=share_link
https://drive.google.com/drive/folders/1pHmE9cg_tMcouV1LFJFtbyBJNp7oQo9J?usp=sharing.

<!--For citation and further information please refer to: __Li, B., Zhang, W., Guo, C. et al. Benchmarking spatial and single-cell transcriptomics integration methods for transcript distribution prediction and cell type deconvolution. Nat Methods (2022). https://doi.org/10.1038/s41592-022-01480-9__.-->


